### PR TITLE
Add heartbeats to stats.

### PR DESCRIPTION
**Pull request by tebadi**

### CODE REVIEW

**Comment by tebadi** [comment]
Sorry about all the format changes.  The only addition is the `register_hearbeat()` in `model.py` and a call to the function in `proc.py`.

File: libs/stats/odc/stats/model.py [hunk 1]
```diff
@@ -234,6 +238,17 @@ def extend_if_needed(self, seconds, buffer_seconds: int = 30) -> bool:
             return self.extend(seconds)
         return True
 
+    def register_heartbeat(self):
```
Comment (Kirill888) [inline libs/stats/odc/stats/model.py]: This is not the right place for this. It should be TaskRunner concern, also location of the file should be configurable, we might be running several workers on the same machine.

**Review by Kirill888 — CHANGES_REQUESTED** [review]
Please do formatting change as a separate commit before any other work.

- Move heartbeat into TaskRunner
- Make it configurable with default being switched off
- Add cli option to `odc-stats run` to set heartbeat file location

File: libs/stats/odc/stats/model.py [hunk 2]
```diff
@@ -555,6 +555,9 @@ def default_cog_settings():
     )
     overwrite: bool = False
 
+    # Heartbeat filepath
+    heartbeat_filepath: str = None
```
Comment (Kirill888) [inline libs/stats/odc/stats/model.py]: If it can be None, it should `Optional[str]`

File: libs/stats/odc/stats/proc.py [hunk 3]
```diff
@@ -181,7 +191,7 @@ def _run(self, tasks: Iterable[Task]) -> Iterator[TaskResult]:
                 t0 = tk.start_time
             else:
                 t0 = datetime.utcnow()
-
+
```
Comment (Kirill888) [inline libs/stats/odc/stats/proc.py]: trailing whitespace

File: libs/stats/odc/stats/proc.py [hunk 4]
```diff
@@ -166,6 +166,16 @@ def _safe_result(self, f: Future, task: Task) -> TaskResult:
...
 
+    def _register_heartbeat(self, hearbeat_filepath: str):
+        """
+        Records the timestamp at which a hearbeat was detected
+
+        """
+        t_now = self.now()
+        file_obj = open(f"{hearbeat_filepath}/hearbeat.tmp", "w")
+        file_obj.write(t_now.strftime("%Y-%m-%d %H:%M:%S"))
+        file_obj.close()
```
Comment (Kirill888) [inline libs/stats/odc/stats/proc.py]:
```suggestion
        with open(hearbeat_filepath, "w") as file_obj:
            file_obj.write(t_now.strftime("%Y-%m-%d %H:%M:%S"))
```

File: libs/stats/odc/stats/_cli_run.py [hunk 5]
```diff
@@ -107,11 +113,12 @@ def run(
             filedb=filedb,
             plugin=plugin,
             threads=threads,
-            memory_limit=memory_limit,
+            memory_limit=memory_limit,
```
Comment (Kirill888) [inline libs/stats/odc/stats/_cli_run.py]: whitespace?

**Comment by Kirill888** [comment]
@tebadi can you please rebase into single commit?

Something like


- `git reset --soft HEAD~3` -- this removes last 3 commits, but keeps code changes on disk
- Make further cleanup changes
- `git commit` commit current state as one single change
- `git push --force-with-lease` push to github, needs extra arg as you changed history

File: libs/stats/odc/stats/proc.py (hunk 4 above)
Comment (Kirill888) [reply libs/stats/odc/stats/proc.py]: Hold on, we want actual filepath supplied, not directory for it.

Comment (tebadi) [reply libs/stats/odc/stats/proc.py]: Initially I was thinking of having a file with a known path that I can read in the liveness prob section of the job template.  We can use a variable inside the template and pass it to the cli as well as using it in the liveness check I suppose.

Comment (Kirill888) [reply libs/stats/odc/stats/proc.py]: yes, share path via ENV var between launch command and liveness probe script.

**Review by Kirill888 — APPROVED** [review]
:+1:

**Pull request merged**